eth2near/eth2near-block-relay-rs/src/eth2near_relay.rs

```diff
@@ -62,254 +88,565 @@ impl Eth2NearRelay {
+        match self
+            .beacon_rpc_client
+            .get_block_number_for_slot(types::Slot::new(slot))
+        {
+            Ok(block_number) => {
+                return self
+                    .eth1_rpc_client
+                    .get_block_header_by_number(block_number);
+            }
+            Err(err) => Err(err),
+        }
+    }
 }
 
 // Implementation of functions for submitting light client updates
 impl Eth2NearRelay {
+    fn is_enough_blocks_for_update(
+        &self,
+        last_submitted_slot: u64,
+        last_finalized_slot_on_near: u64,
+        last_finalized_slot_on_eth: u64,
+    ) -> bool {
+        if (last_submitted_slot as i64) - (last_finalized_slot_on_near as i64)
+            < (ONE_EPOCH_IN_SLOTS * self.light_client_updates_submission_frequency_in_epochs) as i64
+        {
+            info!(target: "relay", "Light client update were send less then {} epochs ago. Skipping sending light client update", self.light_client_updates_submission_frequency_in_epochs);
+            return false;
+        }
+
+        if last_finalized_slot_on_eth <= last_finalized_slot_on_near {
+            info!(target: "relay", "Last finalized slot on Eth equal to last finalized slot on NEAR. Skipping sending light client update.");
+            return false;
+        }
+
+        true
+    }
+
     fn send_light_client_updates(&mut self, last_submitted_slot: u64) {
         info!(target: "relay", "= Sending light client update =");
 
-        let finalized_block_hash: H256 = match self
-            .eth_client_contract
-            .get_finalized_beacon_block_hash()
-        {
-            Ok(block_hash) => block_hash,
-            Err(err) => {
-                warn!(target: "relay", "Error \"{}\" on getting finalized block hash. Skipping sending light client update", err);
-                return;
-            }
-        };
+        let last_finalized_slot_on_near: u64 = return_on_fail!(
+            self.eth_client_contract.get_finalized_beacon_block_slot(),
+            "Error on getting finalized block hash. Skipping sending light client update"
+        );
 
-        let last_finalized_slot_on_near: u64 = match self
+        let last_finalized_slot_on_eth: u64 = return_on_fail!(self
             .beacon_rpc_client
-            .get_slot_by_beacon_block_root(finalized_block_hash)
+            .get_last_finalized_slot_number(),
+            "Error on getting last finalized slot number on Ethereum. Skipping sending light client update").as_u64();
+
+        trace!(target: "relay", "last_finalized_slot on near/eth {}/{}", last_finalized_slot_on_near, last_finalized_slot_on_eth);
+
+        if let Some(_) = self.next_light_client_update {
+            info!(target: "relay", "Try sending light client update from file");
+            self.seng_light_client_update_from_file(last_submitted_slot);
+            return;
+        }
```

Comment:
We have an option of sending light client update from file: if file is specified, we will send all blocks till the update, and this specific light client update.

next_light_client_update is Option which contain LightClientUpdate from file if it were specified and None otherwise. May be here it will be more clear: https://github.com/aurora-is-near/rainbow-bridge/pull/787/commits/469731cc1b25c6187cbcccbd117bbc7188fcab4e

Reply:
I think it's quite weird for us as this won't help to synchronize. The idea for a single shot run was to fetch the beacon state from the file and then build the custom light client update as we already built in the code. But in this case, we need to provide a light client update in the file, and it seems as it looks like even if we obtain a beacon state and the relayer is not able to do so (for some reason), then we need somehow to build custom LightClientUpdate without the relayer's code.

Reply:
TBC in #798